tokio/src/task/builder.rs

```diff
@@ -0,0 +1,105 @@
+#![allow(unreachable_pub)]
+use crate::util::error::CONTEXT_MISSING_ERROR;
+use crate::{runtime::context, task::JoinHandle};
+use std::future::Future;
+
+/// Factory which is used to configure the properties of a new task.
+///
+/// Methods can be chained in order to configure it.
+///
+/// Currently, there is only one configuration option:
+///
+/// - [`name`], which specifies an associated name for
+///   the task
+///
+/// There are three types of task that can be spawned from a Builder:
+/// - [`spawn_local`] for executing futures on the current thread
+/// - [`spawn`] for executing [`Send`] futures on the runtime
+/// - [`spawn_blocking`] for executing blocking code in the
+///   blocking thread pool.
+///
+/// ## Example
+///
+/// ```no_run
+/// use tokio::net::{TcpListener, TcpStream};
+///
+/// use std::io;
+///
+/// async fn process(socket: TcpStream) {
+///     // ...
+/// # drop(socket);
+/// }
+///
+/// #[tokio::main]
+/// async fn main() -> io::Result<()> {
+///     let listener = TcpListener::bind("127.0.0.1:8080").await?;
+///
+///     loop {
+///         let (socket, _) = listener.accept().await?;
+///
+///         tokio::task::Builder::new()
+///             .name("tcp connection handler")
+///             .spawn(async move {
+///                 // Process each socket concurrently.
+///                 process(socket).await
+///             });
+///     }
+/// }
+/// ```
+#[derive(Default, Debug)]
+pub struct Builder<'a> {
+    name: Option<&'a str>,
+}
+
+impl Builder<'_> {
+    /// Creates a new task builder.
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Assigns a name to the task which will be spawned.
+    pub fn name<'b>(&self, name: &'b str) -> Builder<'b> {
+        Builder { name: Some(name) }
+    }
```

Comment:
This seems dangerous. If we add another option that also has a lifetime in the future, then this signature will not work because that other option's lifetime will continue to be `'a`, not `'b`. To handle that, we would need two lifetime parameters, but afaik adding a new lifetime parameter is a breaking change.

Reply:
Wouldn't adding a second lifetime to a public struct already be a breaking change?

Reply:
Yes. My point is that you also can't reuse the same lifetime for both fields.

Reply:
What other lifetime are we discussing here, and what is the proposed change? Only accepting `&'static str` (or `Into<Cow<'static, str>>` in order to not lose support for dynamic names, which we established we didn't want to enforce against)? I'm not sure what you're proposing

Reply:
We couldn't add another lifetime once this is stable so I don't understand the concern

Reply:
It's more that this introduces a convention for having borrowed options for the builder. If we add any other string arguments, then we would be in trouble if the convention is to store them as a reference.

Reply:
If the signature is changed to this, then reusing the `'a` for all parameters that are by reference would not be a problem.
```Rust
impl<'a> Builder<'a> {
    /// Assigns a name to the task which will be spawned.
    pub fn name(&self, name: &'a str) -> Builder<'a> {
        Builder { name: Some(name) }
    }
}
``

Reply:
That departs from my understanding of the utility of a builder and from the precedent at [`std::thread::Builder`](https://doc.rust-lang.org/stable/std/thread/struct.Builder.html) — if there are later fields, someone might _only_ want to set the other fields but not the name. I'd be more inclined to eliminate the lifetime with `Into<Cow<'static, str>>` or if that's against tokio convention, `&'static str`. Probably if we go with `&'static str`, it won't inconvenience anyone once we figure out how to add span attributes/fields to the task. @carllerche @hawkw anyone opposed to making this `&'static str` instead of `&'a str`?

Reply:
Thanks for the suggestion, addressed in 183d55635f04f4bd509ef7bc124d569c9da134bd